### PR TITLE
GH-44478: [GLib] Prevent Arrow::ExtensionType.new

### LINKED ISSUE #44478: [GLib] red-arrow(ruby) cause segmentaion fault Arrow::Field.new(:extension, :extension)
## Environment

* Ruby 3.3.3
* Apache Arrow/red-arrow: 17.0.0
* OS: Ubuntu 24.04

## Reproduce steps

```ruby
require 'arrow'
Arrow::Field.new(:extension, :extension)
```

`ruby test.rb`

## Backtrace

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `ruby test.rb'.
Program terminated with signal SIGSEGV, Segmentation fault.
Download failed: Invalid argument.  Continuing without source file ./nptl/./nptl/pthread_kill.c.
#0  __pthread_kill_implementation (no_tid=0, signo=11, threadid=<optimized out>) at ./nptl/pthread_kill.c:44

warning: 44	./nptl/pthread_kill.c: No such file or directory
[Current thread is 1 (Thread 0x741d72168740 (LWP 110934))]
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=11, threadid=<optimized out>) at ./nptl/pthread_kill.c:44
#1  __pthread_kill_internal (signo=11, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  __GI___pthread_kill (threadid=<optimized out>, signo=signo@entry=11) at ./nptl/pthread_kill.c:89
#3  0x0000741d7184526e in __GI_raise (sig=sig@entry=11) at ../sysdeps/posix/raise.c:26
#4  0x0000741d71e55b23 in ruby_default_signal (sig=sig@entry=11) at signal.c:422
#5  0x0000741d71c44393 in rb_bug_for_fatal_signal
    (default_sighandler=0x0, sig=sig@entry=11, ctx=ctx@entry=0x55679cf91d80, fmt=fmt@entry=0x741d71f88409 "Segmentation fault at %p") at error.c:1069
#6  0x0000741d71e54dcf in sigsegv (sig=11, info=0x55679cf91eb0, ctx=0x55679cf91d80) at signal.c:926
#7  0x0000741d71845320 in <signal handler called> () at /lib/x86_64-linux-gnu/libc.so.6
#8  0x0000741d55142682 in garrow_data_type_get_raw(_GArrowDataType*) () at /lib/x86_64-linux-gnu/libarrow-glib.so.1700
#9  0x0000741d55142fba in garrow::GExtensionType::GExtensionType(_GArrowExtensionDataType*) ()
    at /lib/x86_64-linux-gnu/libarrow-glib.so.1700
#10 0x0000741d55142731 in garrow_data_type_get_raw(_GArrowDataType*) () at /lib/x86_64-linux-gnu/libarrow-glib.so.1700
#11 0x0000741d5515f309 in garrow_field_new_full () at /lib/x86_64-linux-gnu/libarrow-glib.so.1700
#12 0x0000741d720ffb16 in ffi_call_unix64 () at ../src/x86/unix64.S:104
#13 0x0000741d720fc3ef in ffi_call_int
    (cif=cif@entry=0x7fff046531e0, fn=<optimized out>, rvalue=<optimized out>, avalue=<optimized out>, closure=closure@entry=0x0)
    at ../src/x86/ffi64.c:673
#14 0x0000741d720ff0be in ffi_call
    (cif=cif@entry=0x7fff046531e0, fn=fn@entry=0x741d5515f2d0 <garrow_field_new_full>, rvalue=rvalue@entry=0x7fff046531c8, avalue=avalue@entry=0x7fff04653100) at ../src/x86/ffi64.c:710
#15 0x0000741d5769759b in g_callable_info_invoke
--Type <RET> for more, q to quit, c to continue without paging--c
    (info=info@entry=0x55679dac1040, function=<optimized out>, in_args=in_args@entry=0x55679dff5450, n_in_args=n_in_args@entry=3, out_args=out_args@entry=0x0, n_out_args=n_out_args@entry=0, return_value=<optimized out>, is_method=<optimized out>, throws=<optimized out>, error=<optimized out>) at ../girepository/gicallableinfo.c:775
#16 0x0000741d576981fe in g_function_info_invoke
    (info=info@entry=0x55679dac1040, in_args=0x55679dff5450, n_in_args=3, out_args=0x0, n_out_args=0, return_value=return_value@entry=0x7fff04653328, error=0x7fff04653308) at ../girepository/gifunctioninfo.c:284
#17 0x0000741d576c1384 in rb_gi_function_info_invoke_raw_call (data=0x7fff04653310) at rb-gi-function-info.c:151
#18 rb_gi_function_info_invoke_raw
    (info=info@entry=0x55679dac1040, rb_info=rb_info@entry=127669202246200, rb_receiver=rb_receiver@entry=4, rb_arguments=<optimized out>, 
    rb_arguments@entry=127669192722080, return_value=return_value@entry=0x7fff046533f8, rb_return_value=rb_return_value@entry=0x0) at rb-gi-function-info.c:205
#19 0x0000741d576bfd24 in rg_invoke (self=127669202246200, rb_receiver=127669192727400, rb_arguments=127669192722080)
    at rb-gi-constructor-info.c:126
#20 0x0000741d71ede8c9 in vm_call_cfunc_with_frame_
    (stack_bottom=<optimized out>, argv=<optimized out>, argc=<optimized out>, calling=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3490
#21 vm_call_cfunc_with_frame (calling=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3518
#22 vm_call_cfunc_other (ec=0x55679cf004e0, reg_cfp=0x741d71b15de0, calling=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3544
#23 0x0000741d71ee5370 in vm_sendish
    (ec=0x55679cf004e0, reg_cfp=0x741d71b15de0, cd=0x55679d2b82b0, block_handler=<optimized out>, method_explorer=mexp_search_method) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_callinfo.h:403
#24 0x0000741d71ef1f30 in vm_exec_core (ec=0x7fff04652e10, ec@entry=0x55679cf004e0)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/insns.def:814
#25 0x0000741d71ef6909 in vm_exec_loop (result=36, tag=<optimized out>, state=<optimized out>, ec=<optimized out>) at vm.c:2513
#26 rb_vm_exec (ec=ec@entry=0x55679cf004e0) at vm.c:2489
#27 0x0000741d71f07e6e in invoke_block
    (captured=0x741d71b15ea0, opt_pc=<optimized out>, type=<optimized out>, cref=0x0, self=127669192727400, iseq=0x741d4d72b490, ec=0x55679cf004e0) at vm.c:1509
#28 invoke_iseq_block_from_c
    (me=0x0, is_lambda=<optimized out>, cref=0x0, passed_block_handler=0, kw_splat=0, argv=0x7fff04653768, argc=1, self=127669192727400, captured=0x741d71b15ea0, ec=0x55679cf004e0) at vm.c:1579
#29 invoke_block_from_c_bh
    (force_blockarg=<optimized out>, is_lambda=<optimized out>, cref=<optimized out>, passed_block_handler=<optimized out>, kw_splat=<optimized out>, argv=<optimized out>, argc=<optimized out>, block_handler=<optimized out>, ec=<optimized out>) at vm.c:1597
#30 vm_yield_with_cref (is_lambda=0, cref=0x0, kw_splat=0, argv=0x7fff04653768, argc=1, ec=0x55679cf004e0) at vm.c:1634
#31 vm_yield (kw_splat=0, argv=0x7fff04653768, argc=1, ec=0x55679cf004e0) at vm.c:1642
#32 rb_yield_0 (argv=0x7fff04653768, argc=1) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:1366
#33 catch_i (tag=<optimized out>, _=<optimized out>, argc=<optimized out>, argv=<optimized out>, blockarg=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:2287
#34 0x0000741d71ee8128 in vm_catch_protect (tag=<optimized out>, 
    tag@entry=127669192722120, func=func@entry=0x741d71f07c40 <catch_i>, data=data@entry=0, stateptr=stateptr@entry=0x7fff04653904, ec=ec@entry=0x55679cf004e0) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:2369
#35 0x0000741d71ee83a0 in rb_catch_obj (t=127669192722120, func=0x741d71f07c40 <catch_i>, data=0)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:2395
#36 0x0000741d71ede56c in vm_call_cfunc_with_frame_
    (stack_bottom=<optimized out>, argv=<optimized out>, argc=0, calling=<optimized out>, reg_cfp=0x741d71b15e88, ec=0x55679cf004e0) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3490
#37 vm_call_cfunc_with_frame (ec=0x55679cf004e0, reg_cfp=0x741d71b15e88, calling=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3518
#38 0x0000741d71ee5370 in vm_sendish
    (ec=0x55679cf004e0, reg_cfp=0x741d71b15e88, cd=0x55679da2cea0, block_handler=<optimized out>, method_explorer=mexp_search_method) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_callinfo.h:403
#39 0x0000741d71ef1f30 in vm_exec_core (ec=0x7fff04652e10, ec@entry=0x55679cf004e0)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/insns.def:814
#40 0x0000741d71ef6909 in vm_exec_loop (result=36, tag=<optimized out>, state=<optimized out>, ec=<optimized out>) at vm.c:2513
#41 rb_vm_exec (ec=0x55679cf004e0) at vm.c:2489
#42 0x0000741d71efbee8 in vm_yield_with_cref (is_lambda=0, cref=0x0, kw_splat=0, argv=0x7fff04653c38, argc=1, ec=<optimized out>)
    at vm.c:1628
#43 vm_yield (kw_splat=0, argv=0x7fff04653c38, argc=1, ec=<optimized out>) at vm.c:1642
#44 rb_yield_0 (argv=0x7fff04653c38, argc=1) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:1366
#45 rb_yield (val=<optimized out>) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:1382
#46 0x0000741d71c4908c in rb_ary_each (ary=127669192723880) at array.c:2538
#47 0x0000741d71ede56c in vm_call_cfunc_with_frame_
    (stack_bottom=<optimized out>, argv=<optimized out>, argc=0, calling=<optimized out>, reg_cfp=0x741d71b15ef8, ec=0x55679cf004e0) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3490
#48 vm_call_cfunc_with_frame (ec=0x55679cf004e0, reg_cfp=0x741d71b15ef8, calling=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3518
#49 0x0000741d71ee5370 in vm_sendish
    (ec=0x55679cf004e0, reg_cfp=0x741d71b15ef8, cd=0x55679da0efb0, block_handler=<optimized out>, method_explorer=mexp_search_method) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_callinfo.h:403
#50 0x0000741d71ef1f30 in vm_exec_core (ec=0x7fff04652e10, ec@entry=0x55679cf004e0)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/insns.def:814
#51 0x0000741d71ef6909 in vm_exec_loop (result=36, tag=<optimized out>, state=<optimized out>, ec=<optimized out>) at vm.c:2513
#52 rb_vm_exec (ec=0x55679cf004e0) at vm.c:2489
#53 0x0000741d71efd40a in vm_call0_cc
    (ec=0x55679cf004e0, recv=<optimized out>, id=3137, argc=<optimized out>, argv=<optimized out>, cc=<optimized out>, kw_splat=0) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:110
#54 0x0000741d71eff416 in rb_call (scope=CALL_FCALL, argv=0x741d71a16048, argc=2, mid=3137, recv=127669192727400)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:899
#55 rb_funcallv_kw
    (recv=recv@entry=127669192727400, mid=mid@entry=3137, argc=argc@entry=2, argv=argv@entry=0x741d71a16048, kw_splat=<optimized out>) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_eval.c:1092
#56 0x0000741d71d0a776 in rb_obj_call_init_kw
    (obj=obj@entry=127669192727400, argc=argc@entry=2, argv=argv@entry=0x741d71a16048, kw_splat=<optimized out>) at eval.c:1709
#57 0x0000741d71da5932 in rb_class_new_instance_pass_kw (argc=2, argv=0x741d71a16048, klass=127669341104480) at object.c:2140
#58 0x0000741d71ede8c9 in vm_call_cfunc_with_frame_
    (stack_bottom=<optimized out>, argv=<optimized out>, argc=<optimized out>, calling=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3490
#59 vm_call_cfunc_with_frame (calling=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3518
#60 vm_call_cfunc_other (ec=0x55679cf004e0, reg_cfp=0x741d71b15fa0, calling=<optimized out>)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_insnhelper.c:3544
#61 0x0000741d71ef06b5 in vm_sendish
    (method_explorer=<optimized out>, block_handler=<optimized out>, cd=<optimized out>, reg_cfp=<optimized out>, ec=<optimized out>) at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/vm_callinfo.h:403
#62 vm_exec_core (ec=0x7fff04652e10, ec@entry=0x55679cf004e0)
    at /tmp/ruby-build.20240629234751.4521.fSVHVV/ruby-3.3.3/insns.def:834
#63 0x0000741d71ef6986 in vm_exec_loop (result=36, tag=0x7fff04654270, state=RUBY_TAG_RETURN, ec=0x55679cf004e0) at vm.c:2513
#64 rb_vm_exec (ec=0x55679cf004e0) at vm.c:2492
#65 0x0000741d71ef784b in rb_iseq_eval_main (iseq=<optimized out>) at vm.c:2752
#66 0x0000741d71d06975 in rb_ec_exec_node (ec=ec@entry=0x55679cf004e0, n=n@entry=0x741d7169d6d0) at eval.c:287
#67 0x0000741d71d0a693 in ruby_run_node (n=0x741d7169d6d0) at eval.c:328
#68 0x000055679c6b3186 in rb_main (argv=0x7fff04654538, argc=2) at ./main.c:39
#69 main (argc=<optimized out>, argv=<optimized out>) at ./main.c:58
```

```
-- Ruby level backtrace information ----------------------------------------
test.rb:2:in `<main>'
test.rb:2:in `new'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/red-arrow-17.0.0/lib/arrow/field.rb:156:in `initialize'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:364:in `initialize'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:364:in `each'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:365:in `block in initialize'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:365:in `catch'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:366:in `block (2 levels) in initialize'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:715:in `invoke'
/home/user/.rbenv/versions/3.3.3/lib/ruby/gems/3.3.0/gems/gobject-introspection-4.2.4/lib/gobject-introspection/loader.rb:715:in `invoke'

``` 

## Additional info

Execution log
[err.log](https://github.com/user-attachments/files/17447792/err.log)

### Component(s)

GLib

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
### Rationale for this change

This change fixes GH-44478

### What changes are included in this PR?

This change Prevent `Arrow::ExtensionType.new`. This change fixes segmentation fault.

### Are these changes tested?

YES

Before this change: Cause segmentation fault.
After this change: raise the exception(TypeError)

### Are there any user-facing changes?

N/A

Co-Authored-by: Sutou Kouhei <kou@clear-code.com>
* GitHub Issue: #44478

### CODE REVIEW

File: c_glib/test/test-extension-data-type.rb [hunk 1]
```diff
@@ -102,4 +102,11 @@ def test_wrap_chunked_array
                    extension_chunked_array.chunks.collect(&:class),
                  ])
   end
+
+  # GH-44478 test https://github.com/apache/arrow/issues/44478
```
Comment (kou) [inline c_glib/test/test-extension-data-type.rb:106]: We don't need this comment.

File: c_glib/test/test-extension-data-type.rb [hunk 2]
```diff
@@ -102,4 +102,11 @@ def test_wrap_chunked_array
...
+  def test_abstract_class_init
```
Comment (kou) [inline c_glib/test/test-extension-data-type.rb:107]:
```suggestion
  def test_abstract_class
```

File: c_glib/test/test-extension-data-type.rb [hunk 3]
```diff
@@ -102,4 +102,11 @@ def test_wrap_chunked_array
...
+    assert_raise(TypeError) do
+      Arrow::ExtensionDataType.new.name
```
Comment (kou) [inline c_glib/test/test-extension-data-type.rb:109]: Do we need this `.name`?

```suggestion
      Arrow::ExtensionDataType.new
```

Comment (hiroyuki-sato) [reply c_glib/test/test-extension-data-type.rb:109]: * `Arrow::ExtensionDataType.new`: Does not segmentation fault.
* `Arrow::ExtensionDataType.new.name`:  segmentation fault.

That's why I added `.name`.

Comment (kou) [reply c_glib/test/test-extension-data-type.rb:108]: OK. But the new implementation raises by `Arrow::ExtensionDataType.new` and `.name` is never called, right?
If so, we don't need `.name` here.

Comment (hiroyuki-sato) [reply c_glib/test/test-extension-data-type.rb]: Thanks. You are right the new implementation raise the exception by `Arrow::ExtensionDataType.new`. I removed `.name`.

**Review by kou — APPROVED** [review]
+1

**Pull request merged**